File: backend/src/models/user.model.ts
import { EMAIL_REGEX, INTERNATIONAL_PHONE_REGEX, VIETNAM_PHONE_REGEX } from '@/constants/regex';
import { IUser } from '@/types';
import { IAddresses, IHealthProfile, IPreferences, Role, UserStatus, UserTier } from '@/types/user.type';
import { compareValue, hashValue } from '@/utils/bcrypt';
import mongoose from 'mongoose';
import { randomBytes } from 'crypto';

const isValidPhone = (v: string) => VIETNAM_PHONE_REGEX.test(v) || INTERNATIONAL_PHONE_REGEX.test(v);

const AddressSchema = new mongoose.Schema<IAddresses>(
  {
    label: { type: String, required: true, trim: true },
    receiverName: { type: String, required: true, trim: true },
    phone: {
      type: String,
      trim: true,
      validate: {
        validator: function (v: string) {
          if (v == null || v === '') return true;
          return isValidPhone(v);
        },
        message: 'Số điện thoại không hợp lệ',
      },
    },
    detail: { type: String, required: true, trim: true },
    ward: { type: String, required: true, trim: true },
    district: { type: String, trim: true },
    city: { type: String, required: true, trim: true },
    isDefault: { type: Boolean, default: false },
  },
  {
    _id: false,
  }
);

const HealthSchema = new mongoose.Schema<IHealthProfile>(
  {
    allergies: { type: [String], default: [] },
    calories: { type: Number, default: 0 },
  },
  { _id: false }
);

const PreferencesSchema = new mongoose.Schema<IPreferences>(
  {
    dietary: { type: [String], default: [] },
    allergies: { type: [String], default: [] },
    healthGoals: { type: [String], default: [] },
  },
  { _id: false }
);

const UserSchema = new mongoose.Schema<IUser>(
  {
    fullName: { type: String, trim: true },
    username: { type: String, required: true, trim: true },
    email: { type: String, required: true, trim: true, match: EMAIL_REGEX },
    phone: {
      type: String,
      trim: true,
      validate: {
        validator: function (v: string) {
          if (v == null || v === '') return true;
          return isValidPhone(v);
        },
        message: 'Số điện thoại không hợp lệ',
      },
    },
    avatar: { type: String, default: null },
    avatarPublicId: { type: String, default: null },
    passwordHash: { type: String, required: true, minLength: 6 },
    role: { 
      type: String, 
      required: true, 
      enum: [...Object.values(Role), 'ADMIN', 'MANAGER', 'STAFF', 'CUSTOMER'], 
      default: Role.CUSTOMER 
    },
    isHealthSetup: { type: Boolean, default: false },
    loginFailedCount: { type: Number, default: 0 },
    lockedUntil: { type: Date, default: null },
    verifiedAt: { type: Date, default: null },
    status: { 
      type: String, 
      required: true, 
      enum: Object.values(UserStatus), 
      default: UserStatus.ACTIVE 
    },
    collectedPoints: {
      type: Number,
      default: 0,
      min: [0, 'Collected points cannot be negative'],
    },
    accumulatedPoints: {
      type: Number,
      default: 0,
      min: [0, 'Accumulated points cannot be negative'],
    },
    tier: {
      type: String,
      enum: Object.values(UserTier),
      default: UserTier.BRONZE,
    },
    addresses: {
      type: [AddressSchema],
      default: [],
    },
    health: {
      type: HealthSchema,
      default: () => ({ allergies: [] as string[], calories: 0 }),
    },
    preferences: {
      type: PreferencesSchema,
      default: () => ({ dietary: [] as string[], allergies: [] as string[], healthGoals: [] as string[] }),
    },
    receiveCampaignNotifications: {
      type: Boolean,
      default: true,
    },
    reviewModeration: {
      type: {
        toxicCount: { type: Number, default: 0, min: 0 },
        reviewBannedUntil: { type: Date, default: null },
        lastToxicAt: { type: Date, default: null },
      },
      default: () => ({ toxicCount: 0, reviewBannedUntil: null, lastToxicAt: null }),
    },
    storeId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Store',
      default: null,
    },
    referralCode: {
      type: String,
      unique: true,
      uppercase: true,
    },
    referredBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      default: null,
    },
    aiRecommendationsCache: {
      type: {
        data: { type: mongoose.Schema.Types.Mixed }, // Main AI Recommendations
        safeFoodsData: { type: mongoose.Schema.Types.Mixed }, // Safe Foods AI Insights
        updatedAt: { type: Date }
      },
      default: null,
    },
  },
  {
    timestamps: true,
  }
);

// Indexes
UserSchema.index({ email: 1 }, { unique: true });
UserSchema.index({ phone: 1 }, { unique: true });
UserSchema.index({ username: 1 }, { unique: true });
UserSchema.index({ role: 1 });
UserSchema.index({ status: 1 });
UserSchema.index({ storeId: 1 });

// Middleware "pre-validate"
UserSchema.pre('validate', function (next) {
  if (this.addresses && this.addresses.length > 0) {
    this.addresses = this.addresses.filter(a => a.ward && a.receiverName) as any;
  }
  next();
});

// Middleware "pre-save"
UserSchema.pre('save', async function (next) {
  if (this.isNew && !this.referralCode) {
    this.referralCode = `FOODIE-${randomBytes(4).toString('hex').toUpperCase()}`;
  }

  if (!this.isModified('passwordHash')) return next();

  this.passwordHash = await hashValue(this.passwordHash as string);
  next();
});

// Methods
UserSchema.methods.comparePassword = async function (value: string) {
  return await compareValue(value, this.passwordHash);
};

UserSchema.methods.omitPassword = function () {
  const user = this.toObject();
  delete user.passwordHash;
  return user;
};

const UserModel = mongoose.model<IUser>('User', UserSchema, 'users');

export default UserModel;

File: mobile/lib/core/models/user_model.dart
class UserModel {
  final String id;
  final String username;
  final String email;
  final String? fullName;
  final String? phone;
  final String? avatar;
  final String role;
  final String? storeId;
  final int? collectedPoints;
  final String? tier;
  final double? cancellationRate;
  final int? totalOrders;
  final int? cancelledOrders;
  final List<UserAddressModel>? addresses;
  final UserPreferencesModel? preferences;
  final UserHealthModel? health;
  final DateTime? createdAt;

  UserModel({
    required this.id,
    required this.username,
    required this.email,
    this.fullName,
    this.phone,
    this.avatar,
    required this.role,
    this.storeId,
    this.collectedPoints,
    this.tier,
    this.cancellationRate,
    this.totalOrders,
    this.cancelledOrders,
    this.addresses,
    this.preferences,
    this.health,
    this.createdAt,
  });

  factory UserModel.fromJson(Map<String, dynamic> json) {
    return UserModel(
      id: json['_id'] as String? ?? json['id'] as String? ?? '',
      username: json['username'] as String? ?? '',
      email: json['email'] as String? ?? '',
      fullName: json['fullName'] as String?,
      phone: json['phone'] as String?,
      avatar: json['avatar'] as String?,
      role: json['role'] as String? ?? 'CUSTOMER',
      storeId: json['storeId'] as String?,
      collectedPoints: json['collectedPoints'] as int?,
      tier: json['tier'] as String?,
      cancellationRate: (json['cancellationRate'] as num?)?.toDouble(),
      totalOrders: json['totalOrders'] as int?,
      cancelledOrders: json['cancelledOrders'] as int?,
      addresses: json['addresses'] != null
          ? (json['addresses'] as List)
                .map(
                  (e) => UserAddressModel.fromJson(e as Map<String, dynamic>),
                )
                .toList()
          : null,
      preferences: json['preferences'] != null
          ? UserPreferencesModel.fromJson(
              json['preferences'] as Map<String, dynamic>,
            )
          : null,
      health: json['health'] != null
          ? UserHealthModel.fromJson(json['health'] as Map<String, dynamic>)
          : null,
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'username': username,
      'email': email,
      'fullName': fullName,
      'phone': phone,
      'avatar': avatar,
      'role': role,
      'storeId': storeId,
      'collectedPoints': collectedPoints,
      'tier': tier,
      'cancellationRate': cancellationRate,
      'totalOrders': totalOrders,
      'cancelledOrders': cancelledOrders,
      'addresses': addresses?.map((e) => e.toJson()).toList(),
      'preferences': preferences?.toJson(),
      'health': health?.toJson(),
      'createdAt': createdAt?.toIso8601String(),
    };
  }
}

class UserAddressModel {
  final String? id;
  final String? label;
  final String receiverName;
  final String phone;
  final String detail;
  final String ward;
  final String? district;
  final String city;
  final bool isDefault;

  UserAddressModel({
    this.id,
    this.label,
    required this.receiverName,
    required this.phone,
    required this.detail,
    required this.ward,
    this.district,
    required this.city,
    this.isDefault = false,
  });

  factory UserAddressModel.fromJson(Map<String, dynamic> json) {
    return UserAddressModel(
      id: json['_id'] as String?,
      label: json['label'] as String?,
      receiverName: json['receiverName'] as String? ?? '',
      phone: json['phone'] as String? ?? '',
      detail: json['detail'] as String? ?? '',
      ward: json['ward'] as String? ?? '',
      district: json['district'] as String?,
      city: json['city'] as String? ?? '',
      isDefault: json['isDefault'] as bool? ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'label': label,
      'receiverName': receiverName,
      'phone': phone,
      'detail': detail,
      'ward': ward,
      'district': district,
      'city': city,
      'isDefault': isDefault,
    };
  }
}

class UserPreferencesModel {
  final List<String> allergies;
  final List<String> dietary;
  final List<String> healthGoals;

  UserPreferencesModel({
    this.allergies = const [],
    this.dietary = const [],
    this.healthGoals = const [],
  });

  factory UserPreferencesModel.fromJson(Map<String, dynamic> json) {
    return UserPreferencesModel(
      allergies:
          (json['allergies'] as List?)?.map((e) => e as String).toList() ?? [],
      dietary:
          (json['dietary'] as List?)?.map((e) => e as String).toList() ?? [],
      healthGoals:
          (json['healthGoals'] as List?)?.map((e) => e as String).toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'allergies': allergies,
      'dietary': dietary,
      'healthGoals': healthGoals,
    };
  }
}

class UserHealthModel {
  final List<String> allergies;

  UserHealthModel({this.allergies = const []});

  factory UserHealthModel.fromJson(Map<String, dynamic> json) {
    return UserHealthModel(
      allergies:
          (json['allergies'] as List?)?.map((e) => e as String).toList() ?? [],
    );
  }

  Map<String, dynamic> toJson() {
    return {'allergies': allergies};
  }
}
